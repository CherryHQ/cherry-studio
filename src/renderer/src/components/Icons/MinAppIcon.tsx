import { DEFAULT_MIN_APPS } from '@renderer/config/minapps'
import { MinAppType } from '@renderer/types'
import { FC } from 'react'
import styled from 'styled-components'

interface Props {
  app: MinAppType
  sidebar?: boolean
  size?: number
  style?: React.CSSProperties
}

const MinAppIcon: FC<Props> = ({ app, size = 48, style, sidebar = false }) => {
  // First try to find in DEFAULT_MIN_APPS for predefined styling
  const _app = DEFAULT_MIN_APPS.find((item) => item.id === app.id)

  // If found in DEFAULT_MIN_APPS, use predefined styling
  if (_app) {
    return (
      <Container
        src={_app.logo}
        style={{
          border: _app.bodered ? '0.5px solid var(--color-border)' : 'none',
          width: `${size}px`,
          height: `${size}px`,
          backgroundColor: _app.background,
          ...(sidebar ? {} : app.style),
          ...style
        }}
      />
    )
  }

  // If not found in DEFAULT_MIN_APPS but app has logo, use it (for temporary apps)
  if (app.logo) {
    return (
      <Container
        src={app.logo}
        style={{
          border: 'none',
          width: `${size}px`,
          height: `${size}px`,
          backgroundColor: 'transparent',
          ...(sidebar ? {} : app.style),
          ...style
        }}
      />
    )
  }

  return null
}

const Container = styled.img`
  border-radius: 16px;
  user-select: none;
  -webkit-user-drag: none;
`

export default MinAppIcon
