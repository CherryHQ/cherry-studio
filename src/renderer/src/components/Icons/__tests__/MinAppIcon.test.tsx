import { render } from '@testing-library/react'
import { describe, expect, it, vi } from 'vitest'

import MinAppIcon from '../MinAppIcon'

vi.mock('@renderer/config/minapps', () => ({
  DEFAULT_MIN_APPS: [
    {
      id: 'test-app-1',
      name: 'Test App 1',
      logo: '/test-logo-1.png',
      url: 'https://test1.com',
      bodered: true,
      background: '#f0f0f0'
    },
    {
      id: 'test-app-2',
      name: 'Test App 2',
      logo: '/test-logo-2.png',
      url: 'https://test2.com',
      bodered: false,
      background: undefined
    }
  ]
}))

describe('MinAppIcon', () => {
  const mockApp = {
    id: 'test-app-1',
    name: 'Test App',
    url: 'https://test.com',
    style: {
      opacity: 0.8,
      transform: 'scale(1.1)'
    }
  }

  it('should render with default props', () => {
    const { container } = render(<MinAppIcon app={mockApp} />)
    const img = container.querySelector('img')

    expect(img).toBeInTheDocument()
    expect(img).toHaveAttribute('src', '/test-logo-1.png')
    expect(img).toHaveStyle({
      width: '48px',
      height: '48px',
      border: '0.5px solid var(--color-border)',
      backgroundColor: '#f0f0f0',
      opacity: '0.8',
      transform: 'scale(1.1)'
    })
  })

  it('should render with custom size', () => {
    const { container } = render(<MinAppIcon app={mockApp} size={64} />)
    const img = container.querySelector('img')

    expect(img).toHaveStyle({
      width: '64px',
      height: '64px'
    })
  })

  it('should not apply app.style when sidebar is true', () => {
    const { container } = render(<MinAppIcon app={mockApp} sidebar={true} />)
    const img = container.querySelector('img')

    expect(img).not.toHaveStyle({
      opacity: '0.8',
      transform: 'scale(1.1)'
    })
  })

  it('should apply custom style prop', () => {
    const customStyle = { marginTop: '10px', cursor: 'pointer' }
    const { container } = render(<MinAppIcon app={mockApp} style={customStyle} />)
    const img = container.querySelector('img')

    expect(img).toHaveStyle({
      marginTop: '10px',
      cursor: 'pointer'
    })
  })

  it('should render without border when bodered is false', () => {
    const appWithoutBorder = {
      id: 'test-app-2',
      name: 'Test App 2',
      url: 'https://test2.com'
    }
    const { container } = render(<MinAppIcon app={appWithoutBorder} />)
    const img = container.querySelector('img')

    expect(img).toHaveStyle({
      border: 'none'
    })
  })

  it('should return null when app is not found in DEFAULT_MIN_APPS', () => {
    const unknownApp = {
      id: 'unknown-app',
      name: 'Unknown App',
      url: 'https://unknown.com'
    }
    const { container } = render(<MinAppIcon app={unknownApp} />)

    expect(container.firstChild).toBeNull()
  })

  it('should match snapshot', () => {
    const { container } = render(<MinAppIcon app={mockApp} />)
    expect(container.firstChild).toMatchSnapshot()
  })
})
