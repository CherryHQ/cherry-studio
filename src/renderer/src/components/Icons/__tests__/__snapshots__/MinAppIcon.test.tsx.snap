// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MinAppIcon > should match snapshot 1`] = `
<img
  src="/test-logo-1.png"
  style="display: block; width: 48px; height: 48px; object-fit: cover; border-radius: 8px; box-shadow: rgba(0, 0, 0, 0.02) 0px 1px 4px; border: 0.5px solid var(--color-border); background-color: rgb(240, 240, 240); opacity: 0.8; transform: scale(1.1);"
/>
`;
